reference/plugins.rst

```diff
@@ -0,0 +1,282 @@
+        output.info("conanfile_path=%s" % conanfile_path)
+        output.info("reference=%s" % reference.full_repr())
+        output.info("package_id=%s" % package_id)
+        output.info("remote.name=%s" % remote.name)
+
+    def pre_download(output, reference, remote, **kwargs):
+        output.info("reference=%s" % reference.full_repr())
+        output.info("remote.name=%s" % remote.name)
+
+    def post_download(output, conanfile_path, reference, remote, **kwargs):
+        output.info("conanfile_path=%s" % conanfile_path)
+        output.info("reference=%s" % reference.full_repr())
+        output.info("remote.name=%s" % remote.name)
+
+    def pre_download_recipe(output, reference, remote, **kwargs):
+        output.info("reference=%s" % reference.full_repr())
+        output.info("remote.name=%s" % remote.name)
+
+    def post_download_recipe(output, conanfile_path, reference, remote, **kwargs):
+        output.info("conanfile_path=%s" % conanfile_path)
+        output.info("reference=%s" % reference.full_repr())
+        output.info("remote.name=%s" % remote.name)
+
+    def pre_download_package(output, conanfile_path, reference, package_id, remote, **kwargs):
+        output.info("conanfile_path=%s" % conanfile_path)
+        output.info("reference=%s" % reference.full_repr())
+        output.info("package_id=%s" % package_id)
+        output.info("remote.name=%s" % remote.name)
+
+    def post_download_package(output, conanfile_path, reference, package_id, remote, **kwargs):
+        output.info("conanfile_path=%s" % conanfile_path)
+        output.info("reference=%s" % reference.full_repr())
+        output.info("package_id=%s" % package_id)
+        output.info("remote.name=%s" % remote.name)
+
+Functions of the plugins are intended to be self-descriptive regarding to the execution of them. For example, the ``pre_package()`` function
+is called just before the ``package()`` method of the recipe is executed.
+
+For download/upload functions, the ``pre_download()``/``pre_upload()`` function is executed first in an
+:command:`conan download`/:command:`conan upload` command. Then **pre** and **post** ``download_recipe()``/``upload_recipe()`` and its
+subsequent **pre**/**post** ``download_package()``/``upload_package()`` if that is the case. Finally the general
+``post_download()``/``post_upload()`` function is called to wrap up the whole execution.
+
+.. important::
+
+    **Pre** and **post** ``download_recipe()``/``download_package()`` are also executed when installing new recipes/packages from remotes
+    using :command:`conan create` or :command:`conan install`.
+
+Function parameters
+--------------------
+
+Here you can find the description for each parameter:
+
+- **output**: :ref:`Output object<conanfile_output>` to print formatted messages during execution with the name of the plugin and the
+  function executed, e.g., ``[PLUGIN - complete_plugin] post_download_package(): This is the remote name: default``.
```

Comment:
Maybe add the conan reference to it?  (and it will invalidate some of my comments above)

Reply:
Plugin functions might not always attached to a reference or recipe, for example in the future we could add a function to be executed right after a Conan command